### PR TITLE
MP Testing

### LINKED ISSUE #2364: Integration with JUnit for Helidon MP
Add support for creating tests simply annotated with `@HelidonTest` that would run in a container as a bean.

### PR DESCRIPTION
**Pull request by tomas-langer**
Signed-off-by: Tomas Langer <tomas.langer@oracle.com>

Resolves #2364 

Support for MP testing allowing the following:

```java
@HelidonTest
class TestAutoDiscovery {
    @Inject
    @ConfigProperty(name="app.greeting")
    private String greeting;

    @Test
    void testGreeting() {
        assertThat(greeting, is("TestHello"));
    }
}
```

Or for more complex tests:

```java
@HelidonTest
@DisableDiscovery
@AddBean(MyBean.class)
@AddExtension(ConfigCdiExtension.class)
@AddConfig(key = "app.greeting", value = "TestHello")
class TestNoDiscovery {
    @Inject
    private MyBean myBean;

    @Test
    void testGreeting() {
        assertThat(myBean, notNullValue());
        assertThat(myBean.greeting(), is("TestHello"));
    }
}
```

### CODE REVIEW

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     * When discovery is enabled, the whole classpath is scanned for bean archives (jar files containing
+     * {@code META-INF/beans.xml}) and all beans and extensions are added automatically.
+     * <p>
+     * When discovery is disabled, CDI would only contain the CDI implementation itself and beans and extensions added
+     * through annotations {@link io.helidon.microprofile.tests.junit5.AddBean} and
+     * {@link io.helidon.microprofile.tests.junit5.AddExtension}
+     *
+     * @return whether to do discovery, defaults to {@code true}
+     */
+    boolean discovery() default true;
```
Comment (danielkec) [inline microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java:49]: Would be nice to add something like `boolean dirtiesContext() default false;` if we want fresh container for every test

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    public <T> T interceptTestClassConstructor(Invocation<T> invocation,
+                                               ReflectiveInvocationContext<Constructor<T>> invocationContext,
+                                               ExtensionContext extensionContext) {
+
+        // we need to replace instantiation with CDI lookup, to properly injection into fields (and constructors)
+        invocation.skip();
+
+        return container.select(invocationContext.getExecutable().getDeclaringClass())
+                .get();
+    }
```
Comment (danielkec) [inline microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: Works when injecting to field, but constructor param is not injected in my case:
```java
@HelidonTest(discovery = false)
@AddBeans({
        @AddBean(MessagingLivenessCheck.class),
        @AddBean(TestMessagingBean.class),
})
@AddExtensions({
        @AddExtension(ConfigCdiExtension.class),
        @AddExtension(ServerCdiExtension.class),
        @AddExtension(JaxRsCdiExtension.class),
        @AddExtension(HealthCdiExtension.class),
        @AddExtension(MessagingCdiExtension.class),
})
public class MessagingHealthTest {

    private WebClient client;

    @Inject
    public MessagingHealthTest(ServerCdiExtension server) {
        //ServerCdiExtension server = CDI.current().select(ServerCdiExtension.class).get();
        client = WebClient.builder()
                .baseUri("http://localhost:" + server.port())
                .addReader(JsonpSupport.reader())
                .build();
    }
```
org.junit.jupiter.api.extension.ParameterResolutionException: No ParameterResolver registered for parameter [io.helidon.microprofile.server.ServerCdiExtension arg0] in constructor [public io.helidon.microprofile.messaging.health.MessagingHealthTest(io.helidon.microprofile.server.ServerCdiExtension)].
```

```

Comment (danielkec) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: Turns out `io.helidon.microprofile.tests.junit5.HelidonJunitExtension#interceptTestClassConstructor` didn't get invoked at all. 
But when implementing `org.junit.jupiter.api.extension.ParameterResolver`
```java
    @Override
    public boolean supportsParameter(final ParameterContext parameterContext, final ExtensionContext extensionContext) throws ParameterResolutionException {
        return !container.select(parameterContext.getParameter().getType()).isUnsatisfied();
    }

    @Override
    public Object resolveParameter(final ParameterContext parameterContext, final ExtensionContext extensionContext) throws ParameterResolutionException {
        return container.select(parameterContext.getParameter().getType()).get();
    }
```

It constructor params get resolved

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java (hunk 1 above)
Comment (spericas) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java:49]: Related question, how do JUnit extensions work if TestA runs before TestB and sets config, extensions and beans. Can TestB run with a fresh environment?

Comment (danielkec) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java:49]: Can be done by injecting beans to the test methods params instead

Comment (tomas-langer) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java:49]: That is quite expensive and complicated. Considering the test class itself is supposed to be a CDI bean, shutting the container and starting it again between tests would be unexpected (as you would need to run the next test method on a different instance of the bean)

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java (hunk 2 above)
Comment (tomas-langer) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: I will look into the constructor injection

Comment (tomas-langer) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: On the other hand people usually do not expect to add constructors to their test classes...

Comment (danielkec) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: But when the test class becomes a bean ... :smile: , there is no other beforeAll invoked after injection

Comment (spericas) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: More of a nice to have than a must have IMO

**Review by spericas — APPROVED** [review]
Nice!

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java (hunk 2 above)
Comment (tomas-langer) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:124]: This seems to be a small "miss-design" in Junit - they obtain argument definitions, argument values, and then their code is ignored, as I use CDI to instantiate the bean (so all of that gets discarded).
I have created a workaround that does not actually select the constructor parameters from CDI and just returns default values (null or primitive type). Added a test as well

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java (hunk 1 above)
Comment (tomas-langer) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java:49]: I will try to do the following and see how complex it is:
Allow `@HelidonTest(resetPerTest="true")`

In such a case:
1. No injection allowed into the class itself or using constructor
2. Annotations would be combined from class level and method levels
3. Each method would run in a separate container
4. Test methods can have `SeContainer` as a parameter

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java (hunk 2 above)
Comment (danielkec) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonJunitExtension.java:318]: Thx for the hint `@PostConstruct` is just the beforeAll I needed
```java
@HelidonTest(resetPerTest = false)
@DisableDiscovery
@AddBeans({
        @AddBean(MessagingLivenessCheck.class),
        @AddBean(TestMessagingBean.class),
})
@AddExtensions({
        @AddExtension(ConfigCdiExtension.class),
        @AddExtension(ServerCdiExtension.class),
        @AddExtension(JaxRsCdiExtension.class),
        @AddExtension(HealthCdiExtension.class),
        @AddExtension(MessagingCdiExtension.class),
})
class MessagingHealthTest {

    private WebClient client;

    @Inject
    ServerCdiExtension server;

    @Inject
    TestMessagingBean bean;

    @PostConstruct
    void beforeAll() {
        client = WebClient.builder()
                .baseUri("http://localhost:" + server.port())
                .addReader(JsonpSupport.reader())
                .build();
    }
```

File: microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java (hunk 1 above)
Comment (danielkec) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java]: Awesome thats exactly it! Works like a charm!
```java
    private WebClient client;
    private TestMessagingBean bean;

    @BeforeEach
    void beforeEach() {
        ServerCdiExtension server = CDI.current().select(ServerCdiExtension.class).get();
        bean = CDI.current().select(TestMessagingBean.class).get();
        client = WebClient.builder()
                .baseUri("http://localhost:" + server.port())
                .addReader(JsonpSupport.reader())
                .build();
    }

    @Test
    void alivenessWithErrorSignal(SeContainer seContainer) {
        TestMessagingBean bean = seContainer.select(TestMessagingBean.class).get();
```

Comment (danielkec) [reply microprofile/tests/junit5/src/main/java/io/helidon/microprofile/tests/junit5/HelidonTest.java]: And its faster!
Container per method coded manually:
image
Container per method with `@HelidonTest(resetPerTest = true)`:
image

File: docs/mp/testing/01_testing.adoc [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+- you can add `SeContainer` as a method parameter of any test method and you will get the current container
+
+== Usage - configuration
+In addition to the `@AddConfig` annotation, you can also use
+ `@Configuration`.
+
+This allows you to do the following:
+
+- when `useExisting` is set to `true`, the configuration will not be changed
+        and current MP configuration will be sued
```
Comment (danielkec) [inline docs/mp/testing/01_testing.adoc:92]: `sued`

**Review by danielkec — APPROVED** [review]

File: docs/mp/testing/01_testing.adoc (hunk 3 above)
Comment (tomas-langer) [reply docs/mp/testing/01_testing.adoc:92]: the court room awaits

**Review by danielkec — APPROVED** [review]

**Review by spericas — APPROVED** [review]

**Pull request merged**